### PR TITLE
 Improve Fee UX: Bidirectional Fee/Rate Input & Better Rebalance Button

### LINKED ISSUE #443: Alternative way of choosing fees when building transactions
Currently, if someone wants a precise fee (X satoshis exactly), while building the transaction the only way to accomplish this is to fiddle with the fee rate using trial-and-error until the precise fee is achieved (e.g. changing the fee rate from 4 sat/vb, to 5, to 4.5, to 4.6, to 4.55, etc).

It would be nice to be be able to enter EITHER a fee rate OR a specific fee in satoshis, and changes to one will cause the other to automatically recalculate.

### PR DESCRIPTION
**Pull request by Legend101Zz**
### Problem

Users reported two major UX issues when building transactions:

1. **No way to specify exact fee amounts**: Users could only set a fee rate and had to use trial-and-error, repeatedly adjusting the rate (4 → 5 → 4.5 → 4.6 sats/vB) until they achieved the desired total fee. This was frustrating when trying to pay exactly X satoshis in fees.

2. **Unintuitive rebalance button**: The output amount rebalance feature used a small gray +/- icon button that users found difficult to discover and unclear in its purpose.

3. **Auto-spend mode limitations**: In wallet auto-spend mode, the fee amount field was completely hidden, preventing any fee customization beyond the rate.

### Changes Made
Screenshot 2026-01-13 at 20 39 03

### Testing

**Manual Testing Checklist:**

- [ ] **Fee Rate → Fee Amount sync**
  - [ ] Enter fee rate with inputs selected → fee amount updates
  - [ ] Enter fee rate without inputs → fee amount stays "0.0000"
  - [ ] Change fee rate → fee amount recalculates

- [ ] **Fee Amount → Fee Rate sync**
  - [ ] Enter fee amount with inputs selected → rate updates
  - [ ] Enter fee amount without inputs → rate stays same
  - [ ] Helper text shows "Effective rate: X sats/vB"

- [ ] **Rebalance button**
  - [ ] Button is visible and prominent (blue, labeled "Rebalance")
  - [ ] Shows correct action ("Increase to..." or "Decrease to...")
  - [ ] Clicking button adjusts output amount correctly
  - [ ] Only appears when balance error exists and is fixable

- [ ] **Auto-spend mode**
  - [ ] Fee amount field is visible
  - [ ] Can enter exact fee amount
  - [ ] Auto coin selection uses exact fee when specified
  - [ ] Error shown if specified fee is too low
  - [ ] Effective fee rate updates correctly

- [ ] **Edge cases**
  - [ ] Empty fee amount → graceful handling
  - [ ] Zero fee rate → no errors
  - [ ] Very large fees → validation works
  - [ ] Add/remove outputs → fee recalculates correctly

**Test Transactions:**
1. Script Explorer (manual mode)
2. Wallet spend (auto-spend mode)
3. Multi-output transactions
4. Transactions requiring change

### Related Issues

Fixes #443 

---

### CODE REVIEW

**Review by bucko13 — CHANGES_REQUESTED** [review]
a couple of console.logs to cleanup. will validate functionality as well.

**Comment by Legend101Zz** [comment]
@bucko13 sorry for the huge delay on this , finally I have fixed all the issues and also I decided not have the fees edit field on auto mode as it was getting far to complicated and logic was being fragile and inaccurate to calculate the fees before hand without knowing the inputs , can you check the UI , I am happy to iterate as we are taking time so any new fresh and better UX would be very nice 

Screenshot 2026-03-17 at 22 59 19

**Comment by Legend101Zz** [comment]
@bucko13 @tomunchained how do these look ? 
Screenshot 2026-03-17 at 23 10 48


Screenshot 2026-03-17 at 23 10 30

**Comment by bucko13** [comment]
Ugh, bummer. Still having e2e failures! https://github.com/caravan-bitcoin/caravan/actions/runs/23207942129/job/67448546087?pr=447

**Comment by bucko13** [comment]
Ah, i think this is a legitimate e2e failure! 

```
    1 failed
      [mutate-wallet] › e2e/tests/mutate/transaction-flow.spec.ts:84:3 › Transaction Creation and Signing › manual coin selection: create, sign, broadcast 
```

File: apps/coordinator/src/components/ScriptExplorer/OutputsForm.jsx [hunk 1]
```diff
@@ -154,19 +157,31 @@ class OutputsForm extends React.Component {
...
+    }
 
-    if (
-      rate === "" ||
-      Number.isNaN(parseFloat(rate, 10)) ||
-      parseFloat(rate, 10) < 1
-    )
-      rate = "0";
-    setFeeRate(rate);
+    setFeeRate(rate === "" ? "0" : rate); // Reducer handles fee calculation and validation
```
Comment (bucko13) [inline apps/coordinator/src/components/ScriptExplorer/OutputsForm.jsx:166]: Can also do:

```
Math.round(num * 100) / 100
```

File: apps/coordinator/src/reducers/transactionReducer.js [hunk 2]
```diff
@@ -168,7 +170,15 @@ function deleteOutput(state, action) {
 }
 
 function updateFeeRate(state, action) {
-  const feeRateString = action.value;
+  let feeRateString = action.value;
+
+  // Limit to 2 decimal places
+  if (feeRateString && feeRateString.includes(".")) {
```
Comment (bucko13) [inline apps/coordinator/src/reducers/transactionReducer.js]: Same 

```
Math.round((num + Number.EPSILON) * 100) / 100
```

(or similar)

**Review by bucko13 — CHANGES_REQUESTED** [review]
couple small questions/comments. Need to fix e2e as well. I'll test functionally next but the screenshots look good

**Comment by bucko13** [comment]
UX Feedback:

I don't love that the up arrow increments by a hundredth of a sat/vbyte when doing the up arrow. that's a very small unit of change that I don't think has much value 

Screenshot 2026-03-17 at 4 48 10 PM

It looks like the fee rate is not updating with the fee amount 

Screenshot 2026-03-17 at 4 51 43 PM
Screenshot 2026-03-17 at 4 51 58 PM
Screenshot 2026-03-17 at 4 52 09 PM

Got some weird UI juping around when I added a change. There was a

Screenshot 2026-03-17 at 4 53 14 PM

And this was funky too. When i changed the fee amount, rebalance was only available on the recipient output, rather than the change one which is where I would expect it 
Screenshot 2026-03-17 at 4 53 38 PM

**Comment by Legend101Zz** [comment]
> Ah, i think this is a legitimate e2e failure!
> 
> ```
>     1 failed
>       [mutate-wallet] › e2e/tests/mutate/transaction-flow.spec.ts:84:3 › Transaction Creation and Signing › manual coin selection: create, sign, broadcast 
> ```

I yes I suspected that as I made changes to the transcation creation flow so would need to make changes to selectors in the Page model in. e2e , glad it's working fine :)

**Comment by Legend101Zz** [comment]
> UX Feedback:
> 
> I don't love that the up arrow increments by a hundredth of a sat/vbyte when doing the up arrow. that's a very small unit of change that I don't think has much value
> 
> Screenshot 2026-03-17 at 4 48 10 PM
> It looks like the fee rate is not updating with the fee amount
> 
> Screenshot 2026-03-17 at 4 51 43 PM Screenshot 2026-03-17 at 4 51 58 PM Screenshot 2026-03-17 at 4 52 09 PM
> Got some weird UI juping around when I added a change. There was a
> 
> Screenshot 2026-03-17 at 4 53 14 PM
> And this was funky too. When i changed the fee amount, rebalance was only available on the recipient output, rather than the change one which is where I would expect it Screenshot 2026-03-17 at 4 53 38 PM

Thank you for all the testing ... I have resolved all these bugs and flaky behaviour on my end , can you please test it out again :)

**Comment by bucko13** [comment]
Sorry for the delay. I will give this one more review ASAP. could you help cleanup the e2e as well when you get a chance?

**Comment by Legend101Zz** [comment]
> Sorry for the delay. I will give this one more review ASAP. could you help cleanup the e2e as well when you get a chance?

yes sure let me fix that :)

**Comment by bucko13** [comment]
Screenshot 2026-05-26 at 11 01 11 AM

Thanks for fixing the tests! Still running into some issues unfortunately. notice that the "Outputs + Fees" section only reflects the total outputs (output + change) but the fee doesn't get included. If this is meant to just reflect total output, then we can rename the field otherwise need to update the calcuation. 

Otherwise the feature is working nicely! (Also caravan-x which i used to test :) ) 
Screenshot 2026-05-26 at 11 03 59 AM

**Comment by tomunchained** [comment]
When I try to test via vercel "view deployment," I'm seeing no sign of the new feature. I load in a wallet, try to build a withdrawal transaction, and there's only the ability to adjust the fee rate without the absolute fee as shown in the screenshots. I asked another user to try it and he had the same experience. @bucko13 what are we doing wrong?

**Comment by bucko13** [comment]
@tomunchained what do you see in the bottom right hand corner for the version? And the URL? https://caravan-coordinator-denrj7idh-unchained.vercel.app/#/wallet

Screenshot 2026-05-29 at 11 10 44 AM

**Comment by jpcummins** [comment]
Here's what I see:

image

**Comment by bucko13** [comment]
@tomunchained @jpcummins this is only exposed in manual coin selection mode. it's a little unintuitive (I ran into the [same confusion in an earlier review](https://github.com/caravan-bitcoin/caravan/pull/447#issuecomment-3781469689)) but that's because bitcoin fees are kind of unintuitive. In auto mode, the coin selection happens based on how much you're sending and your target fee rate. You can't know the absolute fee until your coins have been selected. You could probably design a UX where a user puts in a target, the coins are selected and then adjustments are made, but it adds a lot more complexity. and also from a usability point of view, you don't know how much you should even be paying until you know the size of the transaction (which is dependent on the inputs selected). 

With manual mode, you know both the inputs and outputs at any given time. This makes setting the fee by rate or absolute number trivial. 

Screenshot 2026-05-29 at 11 43 14 AM
Screenshot 2026-05-29 at 11 43 25 AM

**Review by jpcummins — APPROVED** [review]

File: .changeset/famous-forks-build.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"caravan-coordinator": patch
```
Comment (Shadouts) [inline .changeset/famous-forks-build.md:2]: nit: This should be a minor bump since it's a new feature and does change how rebalance looks.

File: apps/coordinator/src/components/ScriptExplorer/OutputEntry.jsx [hunk 4]
```diff
@@ -266,35 +280,61 @@ class OutputEntry extends React.Component {
...
-              placement="top"
-            >
-              <small>
-                <IconButton onClick={this.handleBalance}>
-                  {this.balanceAction() === "Increase" ? (
+        <Grid item xs={2} sx={{ minHeight: 64 }}>
+          {showRebalance && (
+            <Box>
+              <Button
+                data-testid="rebalance-button"
```
Comment (Shadouts) [inline apps/coordinator/src/components/ScriptExplorer/OutputEntry.jsx:287]: nit: `data-testid` props aren't great because they will render into an attribute on the html element.

File: apps/coordinator/src/components/ScriptExplorer/OutputsForm.jsx [hunk 5]
```diff
@@ -235,190 +250,466 @@ class OutputsForm extends React.Component {
...
-                    Estimated Fees
+                    Fee rate
                   </Typography>
+                </Grid>
+                <Grid item xs={4} sm={3}>
                   <TextField
                     fullWidth
-                    name="fee_total"
+                    value={feeRate}
+                    variant="outlined"
```
Comment (Shadouts) [inline apps/coordinator/src/components/ScriptExplorer/OutputsForm.jsx:325]: nit: This input looks disabled to me on Chrome with a gray background. That's confusing as I didn't think it was editable until I clicked it.

**Review by Shadouts — COMMENTED** [review]
The code is okay. 

Some notes on manual testing:
I found a few bad-state bugs with the automatic<->manual transaction toggle that completely break manual mode. It really is only usable if you toggle it on and never switch back. I know this is outside the scope of your change, but I thought I'd comment on it.

**Review by Shadouts — APPROVED** [review]
